### PR TITLE
Do not compress traces and logs batches when sending to old EVP proxy

### PR DESCRIPTION
**Pull request by nikita-tkachenko-datadog**
### Requirements for Contributing to this repository

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix one issue at the time.
* The pull request must update the test suite to demonstrate the changed functionality.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see [CONTRIBUTING](/CONTRIBUTING.md). 

### What does this PR do?

When sending batched logs or traces in Agentful mode, it checks the version of the EVP proxy available in the DD Agent.
If EVP proxy is older than v4, sent requests are not gzipped (since older EVP proxy versions do not forward content-encoding headers).

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogAgentClient.java [hunk 1]
```diff
@@ -369,25 +368,24 @@ private void flushSafely() throws IOException {
...
-        TraceWriteStrategyImpl evpStrategy = new TraceWriteStrategyImpl(Track.WEBHOOK, payloadSender::send);
-        TraceWriteStrategyImpl apmStrategy = new TraceWriteStrategyImpl(Track.APM, this::sendSpansToApm);
-        return new AgentTraceWriteStrategy(evpStrategy, apmStrategy, this::isEvpProxySupported);
-    }
+        Set<String> agentEndpoints = fetchAgentEndpoints();
+        boolean evpProxyExists = agentEndpoints.contains("/evp_proxy/v3/");
+        if (evpProxyExists) {
+            DatadogGlobalConfiguration datadogGlobalDescriptor = DatadogUtilities.getDatadogGlobalDescriptor();
+            String urlParameters = datadogGlobalDescriptor != null ? "?service=" + datadogGlobalDescriptor.getCiInstanceName() : "";
+            String url = String.format("http://%s:%d/evp_proxy/v1/api/v2/webhook/%s", hostname, traceCollectionPort, urlParameters);
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogAgentClient.java:376]: We're checking `v3` and `v4`, but then we're always sending to `v1`. Could we just drop a comment saying that this is intended, as we discussed offline.

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogApiClient.java [hunk 2]
```diff
@@ -319,14 +319,14 @@ private static final class ApiLogWriteStrategy implements LogWriteStrategy {
...
-            JsonPayloadSender<JSONObject> payloadSender = new CompressedBatchSender<>(
+                    "DD-API-KEY", Secret.toString(apiKey));
+            JsonPayloadSender<JSONObject> payloadSender = new BatchSender<>(
                     httpClient,
                     logIntakeUrl,
                     headers,
                     PAYLOAD_SIZE_LIMIT,
-                    Function.identity());
+                    Function.identity(),
+                    true);
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogApiClient.java:329]: Could we create a constant for this `true` variable to improve readability? Like `private static final boolean compressionEnabled` or something similar.

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogApiClient.java [hunk 3]
```diff
@@ -364,9 +364,8 @@ public TraceWriteStrategy createTraceWriteStrategy() {
 
         Map<String, String> headers = Map.of(
             "DD-API-KEY", Secret.toString(apiKey),
-            "DD-CI-PROVIDER-NAME", "jenkins",
-            "Content-Encoding", "gzip");
-        JsonPayloadSender<Payload> payloadSender = new CompressedBatchSender<>(httpClient, url, headers, PAYLOAD_SIZE_LIMIT, p -> p.getJson());
+            "DD-CI-PROVIDER-NAME", "jenkins");
+        JsonPayloadSender<Payload> payloadSender = new BatchSender<>(httpClient, url, headers, PAYLOAD_SIZE_LIMIT, p -> p.getJson(), true);
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogApiClient.java:368]:
```suggestion
        JsonPayloadSender<Payload> payloadSender = new BatchSender<>(httpClient, url, headers, PAYLOAD_SIZE_LIMIT, p -> p.getJson(), compressionEnabled);
```

Similar comment here.

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/BatchSender.java [hunk 4]
```diff
@@ -22,23 +24,31 @@ public class CompressedBatchSender<T> implements JsonPayloadSender<T> {
...
         this.url = url;
-        this.headers = headers;
+        this.headers = new HashMap<>(headers);
         this.batchLimitBytes = batchLimitBytes;
         this.payloadToJson = payloadToJson;
+        this.compress = compress;
+
+        if (compress) {
+            this.headers.put("Content-Encoding", "gzip");
+        }
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/clients/BatchSender.java:44]: Let's add a comment explaining that Datadog Agent previous version X does not support compression in the proxy endpoint.

**Review by drodriguezhdez — COMMENTED** [review]
Overall looks good. I've dropped some comments to improve readability.

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogAgentClient.java (hunk 1 above)
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogAgentClient.java:377]: Added a comment

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogApiClient.java (hunk 2 above)
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogApiClient.java]: Created a named constant

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogApiClient.java (hunk 3 above)
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/clients/DatadogApiClient.java]: Created a named constant

File: src/main/java/org/datadog/jenkins/plugins/datadog/clients/BatchSender.java (hunk 4 above)
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/clients/BatchSender.java:46]: Added a comment

**Review by drodriguezhdez — APPROVED** [review]

**Pull request merged**